fido/Crypto/Fido2/Model/JavaScript/Types.hs

```diff
@@ -13,6 +14,28 @@ import qualified Crypto.Fido2.PublicKey as PublicKey
 import qualified Data.Aeson as Aeson
 import Data.Map (Map)
 import Data.Text (Text)
+import qualified Deriving.Aeson as Aeson
+import GHC.Generics (Generic)
+
+-- | [(spec)](https://www.w3.org/TR/webauthn-2/#dictionary-client-data)
+-- Intermediate type used to extract the JSON structure stored in the
+-- CBOR-encoded [clientDataJSON](https://www.w3.org/TR/webauthn-2/#dom-authenticatorresponse-clientdatajson).
+data ClientDataJSON = ClientDataJSON
+  { typ :: JS.DOMString,
+    challenge :: JS.DOMString,
+    origin :: JS.DOMString,
+    crossOrigin :: Maybe Bool
+    -- TODO
+    -- tokenBinding :: Maybe TokenBinding
+  }
+  deriving (Generic)
+  -- Note: Encoding can NOT be derived automatically, and most likely not even
+  -- be provided correctly with the Aeson.ToJSON class, because it is only a
+  -- JSON-_compatible_ encoding, but it also contains some extra structure
+  -- allowing for verification without a full JSON parser
+  -- See <https://www.w3.org/TR/webauthn-2/#clientdatajson-serialization>
+  deriving (Aeson.FromJSON) via Aeson.CustomJSON '[Aeson.OmitNothingFields, Aeson.FieldLabelModifier (Aeson.Rename "typ" "type")] ClientDataJSON
+  deriving (Aeson.ToJSON) via Aeson.CustomJSON '[Aeson.OmitNothingFields, Aeson.FieldLabelModifier (Aeson.Rename "typ" "type")] ClientDataJSON
```

Comment:
This needs to be addressed. To be strictly spec-compliant we should implement the JSON-serialization how they explain it. I'm not sure if it matters for the tests, but this is part of the library, and as such we should make sure it's correct.

Reply:
I can use standalone deriving to deriving this instance only in the tests. Is that something you would be okay with or should we implement as described?

Reply:
I think a `FIXME` comment explaining this for now would be alright

Btw you can derive both `FromJSON` and `ToJSON` in one with `deriving (Aeson.FromJSON, Aeson.ToJSON) via Aeson.CustomJSON ...`